clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -350,27 +351,38 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
         if (numHeaders == 0)
             return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
 
+        Header[] headers = getHeaders(buffer, numHeaders);
+
+        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+    }
+
+    private static Header[] getHeaders(ByteBuffer buffer, int numHeaders) {
         Header[] headers = new Header[numHeaders];
-        for (int i = 0; i < numHeaders; i++) {
-            int headerKeySize = ByteUtils.readVarint(buffer);
-            if (headerKeySize < 0)
-                throw new InvalidRecordException("Invalid negative header key size " + headerKeySize);
-
-            String headerKey = Utils.utf8(buffer, headerKeySize);
-            buffer.position(buffer.position() + headerKeySize);
-
-            ByteBuffer headerValue = null;
-            int headerValueSize = ByteUtils.readVarint(buffer);
-            if (headerValueSize >= 0) {
-                headerValue = buffer.slice();
-                headerValue.limit(headerValueSize);
-                buffer.position(buffer.position() + headerValueSize);
+        try {
+            for (int i = 0; i < numHeaders; i++) {
+                int headerKeySize = ByteUtils.readVarint(buffer);
+                if (headerKeySize < 0)
+                    throw new InvalidRecordException("Invalid negative header key size " + headerKeySize);
+
+                String headerKey = Utils.utf8(buffer, headerKeySize);
+                buffer.position(buffer.position() + headerKeySize);
+
+                ByteBuffer headerValue = null;
+                int headerValueSize = ByteUtils.readVarint(buffer);
+                if (headerValueSize >= 0) {
+                    headerValue = buffer.slice();
+                    headerValue.limit(headerValueSize);
+                    buffer.position(buffer.position() + headerValueSize);
+                }
+
+                headers[i] = new RecordHeader(headerKey, headerValue);
             }
-
-            headers[i] = new RecordHeader(headerKey, headerValue);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid number of headers declared for the record");
         }
-
-        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+        if (buffer.position() != buffer.limit())
```

Comment:
Yup. Just realized that the buffer is not exclusive for a given record. Will fix.